### PR TITLE
SlotFill: fix a bug with storing stale fillProps

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes a `SlotFill` bug where a slot is sometimes rendered with old stale `fillProps`. It's described and discussed here: https://github.com/WordPress/gutenberg/pull/65907#issuecomment-2461140952. In short, the `updateSlot` method needs to check a `ref` value to determine if it's still the current registered slot. When an old slot is unregistered and new slot is registered, there is a short period of time where both are semi-registered and special care is needed.

~~That's the first commit in this PR. In addition, I'm doing a series of refactorings of `SlotFill` in order to put it in a better shape:~~

### CODE REVIEW

File: packages/components/src/slot-fill/bubbles-virtually/use-slot.ts [hunk 1]
```diff
@@ -21,8 +21,10 @@ export default function useSlot( name: SlotKey ) {
 
 	const api = useMemo(
 		() => ( {
-			updateSlot: ( fillProps: FillProps ) =>
-				registry.updateSlot( name, fillProps ),
+			updateSlot: (
```
Comment (youknowriad) [inline packages/components/src/slot-fill/bubbles-virtually/use-slot.ts:24]: Where is this one used. or is it not used at all? I see the function arguments modified but I don't see where the calling of the function is updated.

File: packages/components/src/slot-fill/bubbles-virtually/slot.tsx [hunk 2]
```diff
@@ -35,29 +35,33 @@ function Slot(
...
 	const ref = useRef< HTMLElement >( null );
 
+	// We don't want to unregister and register the slot whenever
+	// `fillProps` change, which would cause the fill to be re-mounted. Instead,
+	// we can just update the slot (see hook below).
+	// For more context, see https://github.com/WordPress/gutenberg/pull/44403#discussion_r994415973
+	const fillPropsRef = useRef( fillProps );
+	useLayoutEffect( () => {
+		fillPropsRef.current = fillProps;
+	}, [ fillProps ] );
```
Comment (youknowriad) [inline packages/components/src/slot-fill/bubbles-virtually/slot.tsx:54]: In general this is an indication that something should have been defined using the new `useEvent` maybe.

**Review by youknowriad — COMMENTED** [review]
Left some random comments but to be honest, this one is hard to review. I think I'd just to trust the tests and your confidence in this PR :)

**Comment by jsnajdr** [comment]
> this one is hard to review.

OK, I updated the PR to be smaller and more focused. I'll do the bigger refactors separately.

File: packages/components/src/slot-fill/bubbles-virtually/use-slot.ts (hunk 1 above)
Comment (jsnajdr) [reply packages/components/src/slot-fill/bubbles-virtually/use-slot.ts:24]: The `updateSlot` method is used in the `Slot` component. Called on every render, in a effect without dependencies, to make sure that the latest `fillProps` is always synced promptly to the registry.

The bug being fixed is triggered when, for any reason, the `Slot` is rerendered one more time before unmounting. Then `updateSlot` would store a stale `fillProps` value at a time when a new `Slot` is already being mounted. The `unregisterSlot` method already has exactly the same `ref` check.

File: packages/components/src/slot-fill/bubbles-virtually/slot.tsx (hunk 2 above)
Comment (jsnajdr) [reply packages/components/src/slot-fill/bubbles-virtually/slot.tsx:54]: Unfortunately `useEvent` is specialized to support only callbacks. Here we have a value that is a regular object.

File: packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.tsx [hunk 3]
```diff
@@ -34,22 +34,34 @@ function createSlotRegistry(): SlotFillBubblesVirtuallyContext {
...
+
 			// Make sure we're not unregistering a slot registered by another element
 			// See https://github.com/WordPress/gutenberg/pull/19242#issuecomment-590295412
-			if ( slots.get( name )?.ref === ref ) {
-				slots.delete( name );
+			if ( slot.ref !== ref ) {
+				return;
 			}
+
+			slots.delete( name );
```
Comment (jsnajdr) [inline packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.tsx:48]: Here I'm just tidying up the code, no behavior change. `unregisterSlot` and `updateSlot` now do the same `ref` code and the code looks also the same.

File: packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.tsx [hunk 4]
```diff
@@ -69,20 +81,18 @@ function createSlotRegistry(): SlotFillBubblesVirtuallyContext {
...
 
-	const unregisterFill: SlotFillBubblesVirtuallyContext[ 'registerFill' ] = (
-		name,
-		ref
-	) => {
-		const fillsForName = fills.get( name );
-		if ( ! fillsForName ) {
-			return;
-		}
+	const unregisterFill: SlotFillBubblesVirtuallyContext[ 'unregisterFill' ] =
```
Comment (jsnajdr) [inline packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.tsx:84]: Here I'm fixing a typo in the type: changing `registerFill` to `unregisterFill`. Both functions have the same signature, so the bug didn't have any observable consequences, but worth fixing anyway.

File: packages/components/src/slot-fill/bubbles-virtually/slot.tsx [hunk 5]
```diff
@@ -35,29 +35,33 @@ function Slot(
...
-		// `fillProps` change, which would cause the fill to be re-mounted. Instead,
-		// we can just update the slot (see hook below).
-		// For more context, see https://github.com/WordPress/gutenberg/pull/44403#discussion_r994415973
 	}, [ registerSlot, unregisterSlot, name ] );
-	// fillProps may be an update that interacts with the layout, so we
-	// useLayoutEffect.
+
 	useLayoutEffect( () => {
-		registry.updateSlot( name, fillProps );
+		registry.updateSlot( name, ref, fillPropsRef.current );
```
Comment (jsnajdr) [inline packages/components/src/slot-fill/bubbles-virtually/slot.tsx:64]: This is a critical line in the patch, adding a `ref` param to the `updateSlot` call.

File: packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.tsx [hunk 6]
```diff
@@ -34,22 +34,34 @@ function createSlotRegistry(): SlotFillBubblesVirtuallyContext {
...
 		fillProps
 	) => {
 		const slot = slots.get( name );
 		if ( ! slot ) {
 			return;
 		}
 
+		if ( slot.ref !== ref ) {
+			return;
+		}
```
Comment (jsnajdr) [inline packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.tsx:63]: This is second critical line of the patch, adding a check if `ref` matches.

File: packages/components/src/slot-fill/bubbles-virtually/slot.tsx (hunk 2 above)
Comment (youknowriad) [reply packages/components/src/slot-fill/bubbles-virtually/slot.tsx:54]: What I was thinking is that we could do something like

```
const registerSlotEffect = useEvent(() => {
   registerSlot( name, ref, fillProps );
} );

useEffect(() => {
   registerSlotEffect();
}, [ name ] );
```

File: packages/components/src/slot-fill/bubbles-virtually/use-slot.ts (hunk 1 above)
Comment (youknowriad) [reply packages/components/src/slot-fill/bubbles-virtually/use-slot.ts:24]: But this one updateSlot that is returned from useSlot, is not used right?

**Review by youknowriad — APPROVED** [review]

File: packages/components/src/slot-fill/bubbles-virtually/use-slot.ts (hunk 1 above)
Comment (jsnajdr) [reply packages/components/src/slot-fill/bubbles-virtually/use-slot.ts:24]: Oh yes, this is a good observation. The `updateSlot` and `unregisterSlot` functions returned from `useSlot` are not really used. And I think I could get rid of the entire `api` object and return only the `slot`. I'll do that in a followup PR.

File: packages/components/src/slot-fill/bubbles-virtually/slot.tsx (hunk 2 above)
Comment (jsnajdr) [reply packages/components/src/slot-fill/bubbles-virtually/slot.tsx:54]: Yes, if we really wanted to use `useEvent`, it's possible, but it doesn't really improve the code IMO. Also the `fillPropsRef.current` value is used in two effect hooks, which makes `useEvent` even less attractive.

**Review by Mamaduka — APPROVED** [review]
Thank you, @jsnajdr!

P.S. Inline comments were very helpful!

**Pull request merged**